Change stream field in composebox to be a dropdown.

evykassirer opened the pull request:
This PR changes the stream field in the compose box to be a dropdown menu instead of a free-form field. This helps users search for streams more easily and prevents users from ending up with drafts with half-written (or invalid) stream names.

This dropdown was based off of the dropdown in the "move message" modal.

Fixes #11832


**Screenshots and screen captures:**


![Kapture 2023-02-06 at 16 39 58](https://user-images.githubusercontent.com/5634097/217121984-844012f9-d903-4da8-8f23-9251931dc191.gif)

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [ ] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.

evykassirer commented on static/js/compose.js:
```diff
@@ -458,8 +460,8 @@ export function initialize() {
-        const stream_name = $("#stream_message_recipient_stream").val();
-        if (stream_name === undefined) {
+        const stream_name = compose_state.stream_name();
+        if (stream_name === "") {
```
This is handled in #23351

evykassirer commented on static/js/compose_actions.js:
```diff
@@ -262,7 +278,7 @@ export function update_placeholder_text() {
     const opts = {
         message_type: compose_state.get_message_type(),
-        stream: $("#stream_message_recipient_stream").val(),
+        stream: compose_state.stream_name(),
```
This is handled in https://github.com/zulip/zulip/pull/23351

evykassirer commented on static/js/compose_fade.js:
```diff
@@ -27,8 +27,8 @@ export function set_focused_recipient(msg_type) {
-        const stream_name = $("#stream_message_recipient_stream").val();
-        focused_recipient.topic = $("#stream_message_recipient_topic").val();
+        const stream_name = compose_state.stream_name();
+        focused_recipient.topic = compose_state.topic();
```
This is handled in https://github.com/zulip/zulip/pull/23351

evykassirer commented on static/js/compose_state.js:
```diff
@@ -1,6 +1,7 @@
 import $ from "jquery";
 
 import * as compose_pm_pill from "./compose_pm_pill";
+import * as compose_ui from "./compose_ui";
```
everything in this file is handled in https://github.com/zulip/zulip/pull/23351

evykassirer commented on static/js/compose_state.js:
```diff
@@ -36,8 +37,30 @@ function get_or_set(fieldname, keep_leading_whitespace) {
+        }
+        if (newval !== undefined) {
+            $elem.text(newval);
+            compose_ui.on_compose_select_stream_update();
```
except this part -- this needs to be carried over after the other PR lands

evykassirer commented on static/js/compose_state.js:
```diff
@@ -70,8 +93,9 @@ export function focus_in_empty_compose() {
-            return document.activeElement.value === "";
+            return topic() === "";
+        case "compose_select_stream_name":
+            return stream_name() === "";
```
These three lines also need to be carried over

evykassirer commented on static/js/compose_ui.js:
```diff
@@ -413,3 +418,51 @@ export function get_compose_click_target(e) {
+    // and removes the `open` classname... but then it gets added back!
+    // To reproduce: from topic field, tab backwards and enter to open dropdown,
+    // then click on an option with mouse.
+    // Interestingly, this only happens if the following line of code is uncommented
```
🐛 🐜 🐞  ??

evykassirer commented on static/js/compose_actions.js:
```diff
@@ -62,19 +62,35 @@ function get_focus_area(msg_type, opts) {
     }
```
I wasn't totally sure what belonged in `compose_actions` and not `compose_ui` but I took my best guess

timabbott replied:
`compose_ui` is primarily for clicking buttons in the compose box -- most of the code is around mutating the compose box contents when clicking things. It would be a great preparatory project to rename that module to `compose_textarea_ui.js`, with prep commits to move the few functions that don't belong.

I think in `main`, `compose_ui.js` just has a few things that don't belong: `hide_compose_spinner` and `show_compose_spinner`, with their state variable, which perhaps belong in `compose_state.js`, and `compose_ui.get_compose_click_target`, which probably could be moved into a new `compose_popovers.js` (with other code to migrate there later), or maybe `compose_actions.js` if that doesn't cause import cycles.

@N-Shar-ma FYI -- I think it'd be great if either you or @evykassirer does this rename, as I think it'd make this part of the codebase considerably more readable.

(I think `compose.js` vs. `compose_actions.js` will still have a similar problem after this, but disentangling `compose_textarea_ui.js` will help a lot).

evykassirer replied:
Started thread [here](https://chat.zulip.org/#narrow/stream/6-frontend/topic/compose.20file.20names/near/1492637)

evykassirer commented on frontend_tests/node_tests/composebox_typeahead.js:
```diff
@@ -663,6 +663,16 @@ function sorted_names_from(subs) {
 
 test("initialize", ({override, mock_template}) => {
+    // TODO: where's the right place to put this codeblock to share accross files?
+    // Or is it better to just copy it per file?
```
This gets repeated a lot and would be nice to have in some test util file somewhere

evykassirer replied:
I think it's actually better to leave as this, since it's only 4 instances and there isn't much precedent for test lib files

timabbott replied:
I can't see the context for this diff anymore, but in general it's reasonable to add common code to library files if it feels more readable. (Sometimes, just repeating ourselves is actually more readable; certainly that's going to be the case for things like import statements, where you'd like `grep` to help you find what modules import your module).

evykassirer commented on static/js/compose_ui.js:
```diff
@@ -413,3 +418,47 @@ export function get_compose_click_target(e) {
+    );
+    stream_bar.decorate(new_value, $stream_header_colorblock, true);
+    compose.update_on_recipient_change();
+    // TODO(compose-dropdown)
```
https://chat.zulip.org/#narrow/stream/6-frontend/topic/focus.20on.20composebox

timabbott replied:
Looks like this got resolved in the linked thread.

evykassirer commented on frontend_tests/node_tests/compose.js:
```diff
@@ -493,8 +507,8 @@ test_ui("update_fade", ({override}) => {
-        "#stream_message_recipient_stream,#stream_message_recipient_topic,#private_message_recipient";
-    const keyup_handler_func = $(selector).get_on_handler("keyup");
+        "#compose_select_stream_name,#stream_message_recipient_topic,#private_message_recipient";
+    const keyup_handler_func = $(selector).get_on_handler("change");
```
I changed this to `change` since keyboard input isn't relevant for stream names changing anymore (now it's when an item in the dropdown is selected)

timabbott replied:
That seems correct.

evykassirer commented on static/js/compose_actions.js:
```diff
@@ -62,19 +62,35 @@ function get_focus_area(msg_type, opts) {
 
     if (msg_type === "stream") {
-        return "#stream_message_recipient_stream";
+        return "#id_compose_select_stream";
```
I don't really know how to handle this and it feels a bit janky right now, `#id_compose_select_stream` is the textbox in the dropdown menu, but it's not actually used for anything here.

Essentially, for the stream case instead of focusing we want to click the dropdown instead of doing a focus. 

Is there a nicer way to represent that here?

amanagr replied:
Do we want the dropdown to be closed or open when user clicks `new stream message`? I am sure I understand why you mean by you want it to be clicked and not focused?

alya replied:
Hm, I'm not sure whether we want it to be open. Might be something to try out both ways and see what feels betters.

timabbott replied:
> Soon we'll be using this same dropdown for PMs so we'll probably be removing get_focus_area anyways.

I posted on this here: https://github.com/zulip/zulip/pull/23355#discussion_r1080705508, but I think given this is called in one place, I think it's potentially reasonable to have this just return a string encoding of the "stream select element" as the answer for what should be focused, and have the caller execute "focusing that element" with something more complicated than calling `.focus()`, either via the focus handler on the input doing "open the dropup and focus it" or something else.

But I've not looked at what feels nice closely.

evykassirer commented:
There are two main things I have left to fix -- privacy icons and unit tests. It would be helpful if @timabbott could take a quick lookover what I'm doing to see if I'm moving in the right direction, and I'm happy to clean some things up if that makes it easier to give feedback.

timabbott commented on static/templates/settings/dropdown_list_widget.hbs:
```diff
@@ -6,8 +6,23 @@
+                <i class="fa fa-chevron-up"></i>
+            {{else}}
+                <i class="fa fa-chevron-down"></i>
+            {{/if}}
```
I think this probably live in a separate template; from an abstractions standpoint, `compose` shouldn't appear in the `dropdown_list_widget` codebase at all.

evykassirer replied:
I totally agree! Just gotta wrap my head around how to set that up. I remember we were talking about it on Zulip a while ago.

timabbott commented:
@evykassirer do you have time to update this? It looks like it merge conflicts with your other compose refactoring PRs that we integrated late last year.

evykassirer commented on frontend_tests/node_tests/compose.js:
```diff
@@ -100,6 +100,15 @@ function test_ui(label, f) {
     run_test(label, f);
 }
 
+function mock_stream_header_colorblock() {
```
This function appears in three files -- where's a good shared place I could put it?

timabbott replied:
I think you can make a `compose_lib` file in `frontend_tests/node_tests/lib/` for this.

evykassirer replied:
Great, did that!

evykassirer commented on frontend_tests/node_tests/compose_fade.js:
```diff
@@ -9,7 +9,7 @@ mock_jquery((selector) => {
         case "#stream_message_recipient_stream":
             return {
-                val() {
+                text() {
```
this change happened because the stream input is no longer an input element

evykassirer commented on frontend_tests/node_tests/composebox_typeahead.js:
```diff
@@ -47,7 +47,6 @@ const compose_ui = zrequire("compose_ui");
 const composebox_typeahead = zrequire("composebox_typeahead");
 const settings_config = zrequire("settings_config");
 const pygments_data = zrequire("../generated/pygments_data.json");
-
```
(i'll fix this in cleanup)

evykassirer commented on frontend_tests/node_tests/drafts.js:
```diff
@@ -12,8 +12,9 @@ const compose_pm_pill = zrequire("compose_pm_pill");
-const sub_store = zrequire("sub_store");
+const compose_fade = zrequire("compose_fade");
 const stream_data = zrequire("stream_data");
+const sub_store = zrequire("sub_store");
```
I'll clean this up on an upcoming pass, it got moved for some reason.

evykassirer commented on static/js/click_handlers.js:
```diff
@@ -988,13 +993,16 @@ export function initialize() {
+            const $closest_links = $(e.target).closest("a");
             if (
-                $(e.target).closest("a").length > 0 ||
+                ($closest_links.length > 0 && $closest_links[0].role !== "menuitem") ||
```
Narrowing the scope for this codepath to not have dropdown menu clicks end up here.

timabbott replied:
I would expect this to belong in the `!$(e.target).closest(".overlay").length` class of block just below.

You could probably submit a version of this as a prep commit -- I think basically we just want "clicks inside a `dropdown_list_widget`" never trigger the below logic, just like clicks inside a popover do.

evykassirer replied:
Can you clarify what you mean by belonging in the block below?

`$(e.target).closest("a").length > 0` is true for dropdown clicks, and we need some way to not get inside this if block.

evykassirer replied:
Put a prep commit for this in #24133

evykassirer commented on static/js/compose_actions.js:
```diff
@@ -64,19 +64,35 @@ function get_focus_area(msg_type, opts) {
 export const _get_focus_area = get_focus_area;
 
+export function open_compose_stream_dropup() {
+    const $button = $("#id_compose_select_stream button");
```
I'll add something to exit early if it's already open so that this function can't do a close

evykassirer commented on static/js/compose_actions.js:
```diff
@@ -64,19 +64,35 @@ function get_focus_area(msg_type, opts) {
+        } else {
+            const $elt = $(focus_area);
+            $elt.trigger("focus").trigger("select");
+        }
```
Something I could do instead

```
export function set_focus(msg_type, opts) {
    if (window.getSelection().toString() === "" || opts.trigger !== "message click") {
        if (msg_type === "stream" && !opts.stream) {
            // Open the dropup menu instead of focusing on a field.
            open_compose_stream_dropup();
        } else {
            const focus_area = get_focus_area(msg_type, opts);
            const $elt = $(focus_area);
            $elt.trigger("focus").trigger("select");
        }
```

timabbott replied:
The blueslip error is probably useful; I can easily see a bug where `get_focus_area` no longer works. 

Another idea, which might be cleaner, would be to make `get_focus_area` return a special value for the stream input ... or perhaps best, to write a `.select` handler that opens when that element is focused, and have this code remain unchanged.

As a side note, I noticed that using `tab` and `shift+tab` to navigate between widgets skips the stream input; I'm not entirely sure whether that issue is in this code, but it might be fixed by that last suggestion.

evykassirer replied:
I like that idea! I added a select handler. I'm not seeing the `tab` issue now, but let me know if you still have the issue.

evykassirer commented on static/js/compose_state.js:
```diff
@@ -37,13 +38,18 @@ function get_or_set(fieldname, keep_leading_whitespace) {
 
 export function stream_name() {
-    return $("#stream_message_recipient_stream").val().trim();
+    let val = $("#stream_message_recipient_stream").text().trim();
```
switched from `val()` to `text()` because it's no longer an input field

evykassirer commented on static/js/compose_state.js:
```diff
@@ -37,13 +38,18 @@ function get_or_set(fieldname, keep_leading_whitespace) {
 export function stream_name() {
-    return $("#stream_message_recipient_stream").val().trim();
+    let val = $("#stream_message_recipient_stream").text().trim();
+    if (val === "Select a stream") {
```
It feels a bit messy to have the value be this in the UI and also be `""` in the layer below that, but I couldn't find a nicer way to do this (I forget why at this point I can look into it more).

evykassirer replied:
Ooo one possible solution is `compose_stream_widget.value()` -- I will try that.

evykassirer replied:
Huzzah! That works :)

evykassirer commented on static/js/compose_ui.js:
```diff
@@ -426,3 +431,44 @@ export function get_compose_click_target(e) {
+        widget_name: "compose_select_stream",
+        data: streams_list,
+        default_text: $t({defaultMessage: "Select a stream"}),
+        value: null, // The value is set when the compose box opens.
```
Related to my [other comment](https://github.com/zulip/zulip/pull/23355/commits/69566f08ed6bdae6d3ee738172e1c211a5ba72d6#r1063917832), here it's split into `default_text` and `value`, but once it's rendered I couldn't figure out how to get the pure value again.

evykassirer commented on static/js/composebox_typeahead.js:
```diff
@@ -244,6 +249,11 @@ function handle_keydown(e) {
             // We are doing the focusing on keyup to not abort the typeahead.
             if (on_stream) {
+                if (key === "Enter") {
+                    // This is just opening the dropup menu
```
TODO: verify that this doesn't also capture closing the dropup menu and explain that here

evykassirer commented:
Update from Jan 6:

> This PR feels like a massive change and I'm wondering if I'm missing something that would make it easier. I'll note again that if anyone was able to take a look with me (I'm also happy to talk about it on a call) that would help me feel more confident moving forward. It's a bit hard to look at because of how much I've done so far, but I haven't yet thought of a straightforward way to break it up into functional commits.

I'm going to take a break from working on this PR while I wait for feedback.

timabbott commented on static/js/compose_state.js:
```diff
@@ -37,13 +38,25 @@ function get_or_set(fieldname, keep_leading_whitespace) {
+            return;
+        }
+        $dropdown_button.text(newval);
+        compose_ui.on_compose_select_stream_update();
```
Importing `compose_ui` here reintroduces an import cycle that I think will prevent us from migrating this module to TypeScript.

Probably the simplest answer is to have the callers of this function be responsible for also calling `compose_ui.on_compose_select_stream_update`, and document that in a comment here.

(This could possibly be done as a preparatory refactor commit, with `compose_ui.on_compose_select_stream_update` being basically a noop function).

The other reasonable option would be to interface with this widget via a new module like we do with `compose_pm_pill`, but I'm not sure if that'll actually avoid an import cycle problem.

evykassirer replied:
It feels complicated / confusing to those working in this code in the future, to have callers of this function be responsible for calling the other update functions.

How many import cycles are already present in the code? Could it make more sense to address this issue when thinking about how we want to avoid import cycles more generally?

andersk replied:
Our current import cycle problem is [gigantic](https://chat.zulip.org/#narrow/stream/6-frontend/topic/cyclical.20dependencies/near/1140366) and not getting better fast enough. On the other hand, import cycles do cause real bugs (examples: [1](https://chat.zulip.org/#narrow/stream/6-frontend/topic/all.20messages/near/1147296), [2](https://chat.zulip.org/#narrow/stream/6-frontend/topic/cyclical.20dependencies/near/1165168), [3](https://chat.zulip.org/#narrow/stream/9-issues/topic/.E2.9C.94.20error.20in.20http.3A.2F.2Flocalhost.3A9991.2Fdevtools.2Fintegrations.2F/near/1285725), [4](https://chat.zulip.org/#narrow/stream/6-frontend/topic/Importing.20user_status.20in.20people.2Ejs.20causes.20invalid.20CSRF.20token/near/1326930)) in addition to the maintainability headaches, and we should try pretty hard to avoid making the problem worse.

timabbott commented on static/js/compose_banner.ts:
```diff
@@ -80,5 +80,13 @@ export function show_stream_does_not_exist_error(stream_name: string): void {
     hide_compose_spinner();
-    $("#stream_message_recipient_stream").trigger("focus").trigger("select");
+
+    // TODO: Replace with compose_actions.open_compose_stream_dropup() when it is converted to ts.
```
Hmm, I thought we could call JS code from TS code, it just won't usefully type-check?

evykassirer replied:
I get this error:

```
tsc       | static/js/compose_banner.ts(6,34): error TS7016: Could not find a declaration file for module './compose_actions'. '/srv/zulip/static/js/compose_actions.js' implicitly has an 'any' type.
```

andersk replied:
There are escape hatches, one of which would be to create `static/js/compose_actions.d.ts`:

```ts
export function open_compose_stream_dropup(): void;
```

timabbott commented on static/js/compose.js:
```diff
@@ -405,12 +405,14 @@ export function render_and_show_preview($preview_spinner, $preview_content_box,
-    $(
-        "#stream_message_recipient_stream,#stream_message_recipient_topic,#private_message_recipient",
-    ).on("change", update_on_recipient_change);
+    $("#stream_message_recipient_topic,#private_message_recipient").on(
```
This makes sense -- can you transcribe that observation as a comment for posterity? (Basically just noting the `change` handler suffices for fancier input elements, but we need to listen on `keyup` for plain text fields to see all updates) I can see someone being confused why these two blocks are different.

evykassirer replied:
done!

timabbott commented on frontend_tests/puppeteer_tests/compose.ts:
```diff
@@ -87,10 +93,13 @@ async function test_reply_with_r_shortcut(page: Page): Promise<void> {
+    assert.equal(
+        await common.get_text_from_selector(page, "#compose_select_stream_name"),
+        "Verona",
+    );
```
I would consider doing a prep commit extracting the `common.check_form_contents` block as a function in the puppeteer library file -- something like `await common.check_compose_state(...)` that checks the stream/topic/content in the correct way -- then you end up just needing to do these hunks inside that common function, and we get a more readable test as a bonus.

evykassirer replied:
done in #24133!

timabbott commented on static/js/click_handlers.js:
```diff
@@ -747,6 +747,11 @@ export function initialize() {
+        // The dropdown menu needs to process clicks to open and close.
+        if ($target.parents(".stream_selection_dropdown").length > 0) {
+            return;
+        }
```
I'm a bit puzzled that this click handler would be called before the localized click handlers for the individual widget in the browser's precedence.

evykassirer replied:
We can't prevent default in localized click handlers because the dropdown menu is bootstrap :(

evykassirer replied:
Though hm yeah I'm curious why it went through here before going through bootstrap's handler.

timabbott commented:
Thanks for all the work on this @evykassirer! It continues to be a pretty high priority, since this is the main complex JS/logic change remaining for the compose box redesign.

I think it may be hard to split this into smaller chunks, though I did suggest a handful of preparatory commits/PRs that would help reduce the number of moving parts here.

timabbott commented on static/templates/compose.hbs:
```diff
@@ -65,17 +65,16 @@
                                 <a role="button" class="narrow_to_compose_recipients zulip-icon zulip-icon-arrow-left-circle order-1" data-tooltip-template-id="narrow_to_compose_recipients_tooltip" tabindex="0">
                                 </a>
-                                <input type="text" class="recipient_box" name="stream_message_recipient_stream" id="stream_message_recipient_stream" maxlength="{{ max_stream_name_length }}" value="" placeholder="{{t 'Stream' }}" autocomplete="off" tabindex="0" aria-label="{{t 'Stream' }}" />
+                                <div class="stream_selection_dropdown new-style">
```
What's the thinking on the name `stream_selection_dropdown`? That feels like a good generic name for a reusable "select a stream" component, but this class doesn't apply to similar components in the "move topic" UI, for example. I feel like it'd be nice to have `compose` in the name of the main selector for a component that's only used in the compose box, and I would consider using an ID rather than a class for a component that is conceptually a singleton. (I kinda expect this one will be, because a lot of the selector logic appears related to specific compose box UI details that may not be relevant for similar dropdowns elsewhere).

Also, is the `new-style` CSS class required? If it is, we might want to do a mini prep project to fix that by moving CSS -- I don't think there is a "non-new-style" variant for any of the components inside this div. 

(I think roughly it'll involve making CSS for these components that's currently inside a `new-style` block to just be the CSS for that block, but haven't looked)

evykassirer replied:
Great point; I've updated the classname.

> is the new-style CSS class required?

without it, the dropdown looks like this:

<img width="148" alt="image" src="https://user-images.githubusercontent.com/5634097/213846339-16be1c63-d7f6-4a5a-a384-96c9c704de05.png">

> If it is, we might want to do a mini prep project to fix that by moving CSS -- I don't think there is a "non-new-style" variant for any of the components inside this div.
>
> (I think roughly it'll involve making CSS for these components that's currently inside a new-style block to just be the CSS for that block, but haven't looked)

Can you explain a bit further what you mean by '"non-new-style" variant' and why that's relevant? I think I'm missing some context on what we use `new_style` for.

timabbott replied:
Let me start a #frontend thread about this.

timabbott replied:
https://chat.zulip.org/#narrow/stream/6-frontend/topic/new-style.20CSS.20class/near/1498423

timabbott replied:
For this PR, we can probably just leave `new-style` and hope to remove it in the sweeps described there.

evykassirer commented on frontend_tests/node_tests/composebox_typeahead.js:
```diff
@@ -1121,6 +1093,12 @@ test("initialize", ({override, mock_template}) => {
+    // `if ($(e.target).parents("#compose_select_stream_widget").length > 0) {`
+    // error: "zjquery does not know how to wrap this object yet"
+    // Is there an easy way to mock that out?
+    // Would this whole test be better as an integration test?
```
Question here!

evykassirer replied:
Lucky me, I ended up removing that line so this isn't an issue anymore.

evykassirer commented on frontend_tests/node_tests/composebox_typeahead.js:
```diff
@@ -1209,12 +1186,15 @@ test("initialize", ({override, mock_template}) => {
 
+// I confirmed on main that this test fails if the previous one is removed.
+// Are we okay with unit tests that rely on state from other unit tests?
+/*
```
This test is only failing because I commented out the previous one (which seems nonideal) but I commented it out as well for now.

evykassirer commented on frontend_tests/node_tests/compose.js:
```diff
@@ -293,6 +294,18 @@ test_ui("send_message", ({override, override_rewire, mock_template}) => {
+    mock_stream_header_colorblock();
+    compose.update_on_recipient_change = noop;
+    let stream_value = "";
+    compose_ui.compose_stream_widget = {
```
Needing to do this repeatedly in tests feels a bit messy, not sure if there's a better way to handle this.

evykassirer commented on frontend_tests/node_tests/drafts.js:
```diff
@@ -64,7 +76,6 @@ const timerender = zrequire("timerender");
 
 const draft_1 = {
     stream: "stream",
-    stream_id: 30,
```
I'm not sure why this stuff with `stream_id` 30 got in this diff, note to self to look into it.

evykassirer commented on static/js/hotkey.js:
```diff
@@ -409,8 +409,11 @@ function handle_popover_events(event_name) {
 export function process_enter_key(e) {
-    if ($(".dropdown.open").length && $(e.target).attr("role") === "menuitem") {
-        // on #gear-menu li a[tabindex] elements, force a click and prevent default.
+    if (
```
This is just adding the class `.dropup` because the dropdown added here is technically a dropup menu

evykassirer commented on static/js/compose_state.js:
```diff
@@ -108,8 +112,9 @@ export function focus_in_empty_compose() {
         case "private_message_recipient":
             return private_message_recipient().length === 0;
         case "stream_message_recipient_topic":
-        case "stream_message_recipient_stream":
```
This can be part of a prep commit to use `stream_name()` everywhere

evykassirer commented on static/js/compose.js:
```diff
@@ -484,8 +487,8 @@ export function initialize() {
         (event) => {
             event.preventDefault();
 
-            const stream_name = $("#stream_message_recipient_stream").val();
```
This can be part of a prep commit to use `stream_name()` everywhere

evykassirer commented on frontend_tests/puppeteer_lib/common.ts:
```diff
@@ -186,13 +191,28 @@ export async function check_form_contents(
+    const elements = await page.$$(selector);
+    const texts = await Promise.all(elements.map(async (element) => get_element_text(element)));
+    return texts.join("").trim();
+}
```
I could do a prep commit moving these up

evykassirer commented on static/js/compose_actions.js:
```diff
@@ -63,17 +64,34 @@ function get_focus_area(msg_type, opts) {
+    }
+    $("#compose_select_stream_widget").trigger("focus");
+    $("#id_compose_select_stream").addClass("open");
+    $("#id_compose_select_stream input").trigger("focus");
```
I couldn't get the dropdown() from bootstrap to work (though I didn't spend that long on it), and this feels more clear to me

andersk replied:
`$("#id_compose_select_stream > .dropdown-toggle").dropdown("toggle")` works for me.

evykassirer replied:
🤷 okay cool! thanks :)

evykassirer replied:
switched to that!

evykassirer commented on static/js/composebox_typeahead.js:
```diff
@@ -203,7 +203,6 @@ function handle_keydown(e) {
             target_sel = `#${CSS.escape(e.target.id)}`;
         }
 
-        const on_stream = target_sel === "#stream_message_recipient_stream";
```
streams don't have typeahead anymore

evykassirer commented on static/styles/compose.css:
```diff
@@ -798,6 +783,44 @@ a.compose_control_button.hide {
+    }
+
+    .dropdown-toggle {
+        border-radius: 1px 4px 4px 1px !important;
```
The other `.dropdown-toggle` in the codebase also have this `!important`, but it seems like it might not be necessary. I could look into that separately from this change.

evykassirer commented on static/styles/compose.css:
```diff
@@ -142,13 +142,12 @@
-        left: 6px;
-        top: 0.5px;
+        left: 0;
+        top: 2px;
```
different spacing now that it's inside the dropdown menu

evykassirer commented on static/js/dropdown_list_widget.js:
```diff
@@ -215,6 +217,13 @@ export class DropdownListWidget {
+        if (this.render_pre_dropdown_icon) {
+            const $icon_container = $(`#${CSS.escape(this.container_id)} .pre-dropdown-icon`);
+            if ($icon_container.length) {
+                $icon_container.replaceWith(this.render_pre_dropdown_icon());
```
we could also use `render_stream_privacy` if we're okay with adding the hashtag, and update it from `update_compose_stream_icon`

andersk commented on static/js/compose_state.js:
```diff
@@ -51,13 +52,16 @@ function get_or_set(fieldname, keep_leading_whitespace, no_trim) {
 
 export function stream_name() {
-    return $("#stream_message_recipient_stream").val().trim();
+    if (compose_ui.compose_stream_widget) {
```
Does this ever fail?

evykassirer replied:
Ah, this is initialized once on pageload and not each time the compose box opens. So it would always be defined, cool. Thanks for asking about that.

evykassirer replied:
Removed this check

timabbott commented on static/js/compose_actions.js:
```diff
@@ -63,17 +64,32 @@ function get_focus_area(msg_type, opts) {
-    const focus_area = get_focus_area(msg_type, opts);
     if (window.getSelection().toString() === "" || opts.trigger !== "message click") {
+        const focus_area = get_focus_area(msg_type, opts);
+        if (focus_area === undefined) {
```
I don't think disagree here :)

evykassirer replied:
ah! this is leftover from when I tried changing `get_focus_area` to return undefined for streams. thanks! makes sense

evykassirer replied:
Updated to remove this!

evykassirer commented on static/js/compose_banner.ts:
```diff
@@ -98,5 +98,13 @@ export function show_stream_does_not_exist_error(stream_name: string): void {
+    const $button = $("#id_compose_select_stream button");
+    if ($button) {
+        $button.trigger("click");
+    }
```
oh right I was just telling you about how we do length checks not falsy checks :p 

what's the typescript solution to: `tsc       | static/js/compose_banner.ts(106,55): error TS2339: Property 'dropdown' does not exist on type 'JQuery<HTMLElement>'.`?

evykassirer commented on static/templates/compose.hbs:
```diff
@@ -64,17 +64,16 @@
+                                    {{> settings/dropdown_list_widget
+                                      widget_name="compose_select_stream"
+                                      list_placeholder=(t 'Filter streams')
+                                      dropup="true"}}
```
thanks - changed to that!

evykassirer commented on static/js/hotkey.js:
```diff
@@ -409,8 +409,11 @@ function handle_popover_events(event_name) {
-    if ($(".dropdown.open").length && $(e.target).attr("role") === "menuitem") {
-        // on #gear-menu li a[tabindex] elements, force a click and prevent default.
+    if (
+        ($(".dropdown.open").length || $(".dropup.open").length) &&
```
sweet, thanks! switched to that

alya commented:
Thanks! I think we should ignore capitalization for alphabetization (cf. #23283).

alya commented:
I also noticed that it's buggy in full-sized compose mode:

<img width="526" alt="Screen Shot 2023-02-06 at 11 42 09 PM" src="https://user-images.githubusercontent.com/2090066/217182285-939976e6-a713-4915-affb-c51624a75ce0.png">

alya commented:
I created a new stream to check what happens with long stream names (see "Organization name..." in the left sidebar), but I'm not seeing it showing up in the dropdown.

<img width="919" alt="Screen Shot 2023-02-06 at 11 46 32 PM" src="https://user-images.githubusercontent.com/2090066/217183352-5fb863d6-fb28-4c23-bb6b-3bfb9d22e6a2.png">

alya commented:
A few more questions/thoughts:
1. What do you think about having the angle point down rather than up? It's what we have in the redesign proposal, and the up-angle looks surprising to me (even though I know the menu generally opens upwards).
2. This can be a follow-up, but I think we'll want to show some text when no matches are found (e.g., "No matching streams").
3. Just to check, is adding stream symbols to the typeahead better done as a follow-up?

evykassirer commented:
> I think we should ignore capitalization for alphabetization (cf. https://github.com/zulip/zulip/issues/23283).

done!

> What do you think about having the angle point down rather than up? It's what we have in the redesign proposal, and the up-angle looks surprising to me (even though I know the menu generally opens upwards).

I feel mostly indifferent! I changed it to up.

>  is adding stream symbols to the typeahead better done as a follow-up?

I might be able to add that commit in here if it's easy! Are you thinking of something like this?

<img width="383" alt="image" src="https://user-images.githubusercontent.com/5634097/217675327-c0c39285-36e1-4f82-b1be-07e4b6fa7c90.png">

> This can be a follow-up, but I think we'll want to show some text when no matches are found (e.g., "No matching streams").

I looked around a little bit and couldn't find something that can do this easily, so a follow-up sounds good


looking into the other two issues now

evykassirer commented:
dropdown direction fix:

<img width="407" alt="image" src="https://user-images.githubusercontent.com/5634097/217680257-0b6636bf-34f3-4154-bddf-ee74143a2964.png">

<img width="607" alt="image" src="https://user-images.githubusercontent.com/5634097/217680309-6f779f79-0588-4566-adf3-cc93297898f6.png">

evykassirer commented:
long stream names:

**issue 1**: The dropdown doesn't update until you refresh the page. I don't think there's currently functionality in the dropdown menu to update what options are available, but I can look into a nice way of doing that and hook into that when a new stream is added.

**issue 2**: Long names do break it :p

<img width="626" alt="image" src="https://user-images.githubusercontent.com/5634097/217681494-9fb0891d-d61a-47d9-87b7-0edcbbb67a55.png">

I'm looking into a way of fixing it, will continue on that next time I'm working

alya commented:
> I might be able to add that commit in here if it's easy! Are you thinking of something like this?

I think that's better than not having the `#`! In the end, we want the stream privacy symbols (hash/lock/globe).

alya commented:
> issue 2: Long names do break it :p

I haven't retested just now, but I thought it let me scroll sideways? Though I guess even if it does, that's not immediately obvious.

evykassirer commented:
> The dropdown doesn't update until you refresh the page.

This has been fixed!

> Or are those implemented but just didn't happen to be in the screenshot?

They're not in the dropdown! It's just copying the dropdown from the move-message popup, which also doesn't have the symbols. I think doing this as a followup makes more sense.

Remaining for this PR:

* fix long names

Followups:

* icons in dropdown
* show some text when no matches are found (e.g., "No matching streams")

evykassirer commented:
Added ellipses:

<img width="905" alt="image" src="https://user-images.githubusercontent.com/5634097/218608013-a157269f-5d3e-417d-ac9a-ad7f2efd0f5f.png">

I decided to keep the stream box always the same width, but I'm open to having a different max-width if we want.

alya commented:
> I decided to keep the stream box always the same width, but I'm open to having a different max-width if we want.

I think we should be expanding the box width for longer stream names. We might make use of our horizontal space, and I believe that's the plan for the redesigned version of the compose box as well. (CC @terpimost )

terpimost commented:
have some errors preventing me from testing, shared with @evykassirer in DMs

evykassirer commented:
Sorry about that! Fixed now :)

terpimost commented:
So it is not word wrapping in the list. I thought it should
<img width="444" alt="image" src="https://user-images.githubusercontent.com/1903309/222017518-8796f891-9ed3-4ce2-a03a-86d40ee03991.png">

evykassirer commented:
@terpimost fixed!

<img width="299" alt="image" src="https://user-images.githubusercontent.com/5634097/222257708-6d667424-b9ac-410a-b234-2df947132b8c.png">

<img width="405" alt="image" src="https://user-images.githubusercontent.com/5634097/222257745-f01d6baf-5fb9-4a13-988e-98f29372f23d.png">

alya commented:
> > The dropdown doesn't update until you refresh the page.
> 
> This has been fixed!
> 
> >

Hm, I'm still seeing the dropdown not updating properly. E.g., here I renamed #Scotland:


![Screen Shot 2023-03-03 at 12 48 53 PM](https://user-images.githubusercontent.com/2090066/222826621-2bc5b328-d0f0-481c-b870-91953a40c2d7.png)

The app knows it's not a valid stream name once you select it:

![Screen Shot 2023-03-03 at 12 49 06 PM](https://user-images.githubusercontent.com/2090066/222826634-dae12da4-5b4e-433a-906e-bdd2e05a56ac.png)

evykassirer commented:
Oh fun so these are all different corner cases -- I've handled new channel, subscribe, and unsubscribe -- but rename is a different event handler! Are there others you can think of that are relevant for this kind of thing?

alya commented:
Oh, heh, I see. Archive stream.

alya commented:
Stream privacy changes for the icon of the selected stream (and eventually for the dropdown, once we add icons there).

evykassirer commented:
remaining things to do:

* make width of dropdown menu in the compose box more flexible, from [this comment](https://chat.zulip.org/#narrow/stream/101-design/topic/select.20stream.20dropdown/near/1517433)
* update dropdown options for rename, archive stream, and privacy changes

evykassirer commented:
updates:

* dropdown options update on archive
* dropdown options update on rename
* privacy changes was already working!

I responded to Vlad because I realized he gave me an answer for small screens and I was asking about larger screens, so I'm waiting on that. But otherwise I'm ready for more review @alya !

alya commented:
Thanks! Rename the stream that's currently selected didn't quite work properly:
<img width="178" alt="Screen Shot 2023-03-07 at 12 11 42 AM" src="https://user-images.githubusercontent.com/2090066/223362836-51d24e71-d5ea-4844-bf14-7e32f2356da8.png">

I think the stream should remain selected, under its new name?

alya commented:
I don't see a push, and encountering live updated issues. Did your updates get pushed properly?

evykassirer commented:
no 🤦 sorry, pushed now!

evykassirer commented:
@alya renaming a stream now renames the option in the dropdown menu as well -- ready for your next review :)

evykassirer commented:
I need to fix a failing test but I don't think that's a blocker for product review. I'll get to that next work session.

alya commented:
Thanks! It looks like the stream privacy icon in the selected option is still not live-updating properly:

<img width="480" alt="Screen Shot 2023-03-14 at 11 58 55 PM" src="https://user-images.githubusercontent.com/2090066/225231104-04292559-a921-4a29-a48d-28baae5fc4d1.png">

alya commented:
After unsubscribing from web-public "Rome", I somehow ended up in this incorrect state:
<img width="390" alt="Screen Shot 2023-03-15 at 12 02 56 AM" src="https://user-images.githubusercontent.com/2090066/225232042-8a8b282d-6391-46d8-9914-a0c94c28e60e.png">

 (I don't have a precise recollection of the steps I took, but I saw this after switching to a different window and coming back.)

evykassirer commented:
@alya Thanks, those were real and easily reproducible bugs! Fixed now. Passing back to you :)

alya commented:
Thanks! Where are we at with this?

> remaining things to do:
> 
> * make width of dropdown menu in the compose box more flexible, from [this comment](https://chat.zulip.org/#narrow/stream/101-design/topic/select.20stream.20dropdown/near/1517433)

alya commented:
I didn't spot any live update issues in this round of testing.

evykassirer commented:
Ah yes, the responsiveness is not great -- both for long stream names, and also for small screens actually. I poked at it a bit and was finding it tricky to fix so I asked Anders to help and hopefully will have a fix up soon :)

alya commented:
Ok, sounds good. I think we discussed in #design that we can punt implementing the redesign proposal for small screens (with the topic on a new line) to a follow-up PR.

evykassirer commented:
If you're okay with it breaking smaller than this width, then it's easy to update

<img width="600" alt="image" src="https://user-images.githubusercontent.com/5634097/227376272-a16f52cc-143a-49ec-b94c-3bddf816b9ce.png">

I'm pushing a commit that lets it be this big for bigger names

<img width="897" alt="image" src="https://user-images.githubusercontent.com/5634097/227376480-0aef2a94-0f6d-442b-b2a0-c06f7c8995eb.png">

I haven't figured out how to make it responsive (shrink and grow gradually as the screen changes size) yet.

evykassirer commented:
Responsiveness fixed! It looks like this now:

![Kapture 2023-03-28 at 15 30 53](https://user-images.githubusercontent.com/5634097/228381512-96f8c970-a341-43ef-af53-18f207314e75.gif)


@alya back to you :)

alya commented:
Hm, weird things happen in super-short windows. There's probably not a great solution, but I think the typeahead should at least be on top of everything else:

![Screen Shot 2023-03-28 at 4 43 11 PM](https://user-images.githubusercontent.com/2090066/228390958-93daefd8-04de-4362-8e01-5a5add745f31.png)

alya commented:
That's all I've got. :)

evykassirer commented:
Just fixed that :)

timabbott commented:
I merged 3 prep commits as the series ending with e786c3ac98b1bd1d9f02cddbaa1d1ab303c4e482.

A few functional issues:
* If you type something in the "Filter" box and then hit "Enter", it doesn't select the first item, which it probably should. This is I think a general `dropdown_list_widget` issue; it might make sense to just do a separate mini PR to fix that (once can test with the "Move topic" one I think?).
* Clicking on the grey heading area just above "Filter streams" closes the compose box; I gleaned this from reading the `click_handlers.js` implementation.  `$(e.target).closest("a").attr("role") === "menuitem"` checks if you clicked a menu item, but I think we want to exclude any clicks inside the dropdown container. The fix might be to just add whatever outer class we have on a dropdown_list_widget to the big and gate in that click handler next to `!$(e.target).closest(".popover").length &&`?

timabbott commented on web/src/compose_banner.ts:
```diff
@@ -97,5 +97,10 @@ export function show_stream_does_not_exist_error(stream_name: string): void {
+    if ($("#id_compose_select_stream").hasClass("open")) {
+        return;
+    }
+    $("#id_compose_select_stream button").trigger("click");
```
I don't think we can have this module import `compose_actions.ts` in any case, because that'd generate an import cycle. 

The long term solution is almost certainly that we should be using stream IDs, not names, as the fundamental `compose_state` storage for streams (and probably also the `.value` for the dropdown list widget), which would lead to this error becoming impossible. So I think we can probably leave this as-is but change the TODO concept to be deleting this validator after migrating to stream IDs as how we keep track of the compose box recipient, with the name just used for display.

evykassirer replied:
Great point! Updated the comment.

timabbott commented on web/src/compose_state.js:
```diff
@@ -1,6 +1,7 @@
 import $ from "jquery";
 
 import * as compose_pm_pill from "./compose_pm_pill";
+import * as compose_ui from "./compose_ui";
```
As @andersk noted in a previous review, it is an import cycle to import `compose_ui` from here, so we shouldn't do that.

I'm not sure what the best fix is; one idea would be to have `export let compose_stream_widget` be defined here (but the value populated by a setter inside `compose_ui.js`); another would be to have `on_compose_select_stream_update`, which I think is called whenever that input's value is changed, be responsible for calling `set_stream_name` to update this data structure, and the couple of places that call `set_stream_name` be changed to call a `compose_ui.compose_stream_widget.something` method to actually modify the UI state and thus trigger this data state change?

timabbott commented on web/src/compose_ui.js:
```diff
@@ -2,12 +2,20 @@ import autosize from "autosize";
 import * as common from "./common";
+import * as compose from "./compose";
+import * as compose_actions from "./compose_actions";
+import * as compose_state from "./compose_state";
```
I suspect these new imports are going to generate import cycles as well, hmm. I think the right plan is probably for all your new code here to just be a new module `compose_recipient.js`. We can put both these functions and also the new functions you'd added to `compose_actions`, like `export function close_compose_stream_dropup`, in it; I think we'll be a lot happier with that module structure for other reasons too.

evykassirer replied:
That sounds like a great solution, thanks!

evykassirer replied:
Done and pushed

timabbott commented on web/src/dropdown_list_widget.js:
```diff
@@ -28,7 +29,9 @@ export class DropdownListWidget {
         this.initial_value = value;
+        this.render_pre_dropdown_icon = render_pre_dropdown_icon;
         this.on_update = on_update;
+        this.list_widget = null;
```
Can these changes be a new series of commits, either independently or at the start of the PR? I think it should be easy to extract with `git checkout -p` and an interactive rebase... but I think there's 2 new features here -- the `replace_data` method and `render_pre_dropdown_icon`?

evykassirer replied:
Yeah, that makes sense as a prep commit, thanks! I'm building a model for when to consider something to be a separate commit -- changes to general use modules is one I'll keep an eye on going forward.

evykassirer replied:
Split into its own commit at the start of the PR :) Let me know if you want it in a different PR

timabbott commented on web/src/hotkey.js:
```diff
@@ -402,8 +402,8 @@ function handle_popover_events(event_name) {
-    if ($(".dropdown.open").length && $(e.target).attr("role") === "menuitem") {
-        // on #gear-menu li a[tabindex] elements, force a click and prevent default.
+    if ($(".dropdown.open, .dropup.open").length > 0 && $(e.target).attr("role") === "menuitem") {
+        // on dropdown menu elements, force a click and prevent default.
```
Could this change be an independent prep commit? It seems like a general change to out hotkey handling of `enter` in dropup menus, not something specific to this feature, and it's very helpful to have systematic changes like that be their own commits.

evykassirer replied:
Thanks, done!

timabbott commented on web/src/ui_init.js:
```diff
@@ -632,6 +632,7 @@ export function initialize_everything() {
     user_group_edit.initialize();
     stream_edit_subscribers.initialize();
     stream_data.initialize(stream_data_params);
+    compose_ui.initialize_compose_stream_dropdown();
```
This should probably become `compose_recipient.initialize()` -- our convention is for all of these functions called from this module to be called `initialize`. (It makes it easier to understand modules without every looking at this one).

I also suspect this should be further down -- maybe next to `compose_pm_pill`, where we're initializing the most similar existing UI component?

evykassirer replied:
renamed to `initialize()` and moved next to `compose_pm_pill`

timabbott commented on web/styles/app_components.css:
```diff
@@ -492,6 +492,7 @@ div.overlay {
     @extend .stream-selection-header-colorblock;
     margin-bottom: 5px;
     z-index: 1;
+    min-width: 10px;
```
Is this a bug fix for an issue affecting the move topic modal too? It might be helpful to make this its own commit with an explanation of the bug it's addressing.

evykassirer replied:
It's not a bug fix exactly, but directly related to the rest of this commit, to make sure it doesn't shrink when the stream name overflows. We want to make sure the stream name is cut off before the colorblock shrinks. I can add that to the commit message to be more clear if that helps.

evykassirer replied:
Okay I actually just double checked and I think some CSS fiddling I did since this removed the need for this CSS, so I'll just remove it! Simple solution.

timabbott commented on web/styles/compose.css:
```diff
@@ -774,6 +771,66 @@ a.compose_control_button.hide {
+    }
+
+    .text-warning {
+        color: inherit; /* TODO(compose-dropdown): do we want grey or something here? */
```
I'm not sure how this is displayed, but is this still a TODO we need to resolve?

evykassirer replied:
It's the text color of "Select a stream" here:

<img width="895" alt="image" src="https://user-images.githubusercontent.com/5634097/229023719-760ba06d-2390-4407-aca0-73cdb8e6d307.png">

@alya @terpimost do you have opinions about what color to do here?

alya replied:
The grey seems fine, and since we're getting rid of that bar in the redesign anyway, I think we can reconsider what it looks like when there's no stream selected in the new context.

timabbott replied:
Yeah, that all sounds correct to me.

evykassirer replied:
Removed the comment

evykassirer commented:
> If you type something in the "Filter" box and then hit "Enter", it doesn't select the first item, which it probably should. This is I think a general dropdown_list_widget issue; it might make sense to just do a separate mini PR to fix that (once can test with the "Move topic" one I think?).

Yeah, this isn't specific to this PR. We can make a separate issue for it.

evykassirer commented:
> Clicking on the grey heading area just above "Filter streams" closes the compose box; I gleaned this from reading the `click_handlers.js` implementation.  `$(e.target).closest("a").attr("role") === "menuitem"` checks if you clicked a menu item, but I think we want to exclude any clicks inside the dropdown container. The fix might be to just add whatever outer class we have on a dropdown_list_widget to the big and gate in that click handler next to `!$(e.target).closest(".popover").length &&`?

I didn't add it there because we don't want to do `compose_actions.cancel()`, and we need to do the check before the first `if` which can be true for dropdown clicks, and we don't want to bring focus to the composebox textarea.

That being said, I do think that `if ($(e.target).parents(".dropdown-menu").length) {` is more readable and accurate, so I've changed to that.

_That_ being said, clicking above "Filter streams" doesn't seem to enter that click handler at all? I'm having trouble figuring out what's closing the dropdown menu there.

> `web/templates/compose_stream_dropdown_privacy_icons.hbs` uses a `title` attribute tooltip. I'm not sure we need a tooltip at all there, but if we want one, I would have expected a Tippy tooltip. 

Removed the tooltip for simplicity.

evykassirer commented:
Remaining issues:

- [ ] "Select a stream" color, asked Alya and Vlad. This case doesn't show up in the "move stream" dropdown because a stream is always selected when the dropdown is first visible
- [ ] Clicking on the grey heading area just above "Filter streams" closes the compose box -- I don't feel like this should be blocking, but I can ask in #frontend if you don't have any ideas off the top of your head.

timabbott commented:
>  Clicking on the grey heading area just above "Filter streams" closes the compose box -- I don't feel like this should be blocking, but I can ask in #frontend if you don't have any ideas off the top of your head.

Yeah, I don't want to block on that -- I'm going to try to merge this today.

timabbott commented:
The `compose_recipient` extraction seems helpful but incomplete. I'm going to see if I can make it sane and merge this with several follow-up issues. I opened #25022 for one such issue noticed thus far.

timabbott commented on web/src/compose_recipient.js:
```diff
@@ -0,0 +1,107 @@
+
+import render_compose_stream_dropdown_privacy_icons from "../templates/compose_stream_dropdown_privacy_icons.hbs";
+
+import * as compose from "./compose";
```
We can skip this import if we move a couple functions from `compose.js` here and thus actually avoid introducing the main import cycle; done this in my branch.

There will still be one with `compose_state` and this directly importing each other.

timabbott commented:
OK I've completed the split of `stream_recipient`, to the extent it's easy to do in this pull request, and pushed back here.

Here's the unfortunately long list of outstanding issues as of present. First the UI issues:
* Keyboard UI: If you type something in the filter and hit `Enter`, it should select the first option, without making you hit "Down" first.
* Keyboard UI: If you shift+tab to the stream input, it should open without making you hit `Enter`. As it is, there's no indication that you've tabbed to the right place, which is bad, and also if you start typing once moving focus to the (closed) dropdown_list_widget, it triggers keyboard shortcuts and more or less randomly navigates you to a different part of the app.
* It's way too easy to accidentally close the entire compose box when using this dropdown by trying to click on something -- not just the top grey boundary but anything that isn't a stream.
* Keyboard UI: The underlined focus styling is awkward; this is annoying to debug because trying to use the browser inspector closes everything.
 
Code issues. These are OK as a follow-up, since they have no functional impact, but are definitely problems:
* #25022: Fixing an import cycle involving `input_pill.js` calling a `compose` function. (Not a release blocker).
* Fixing the new simple import cycle between `compose_state` and `compose_recipient` each importing the other.

evykassirer commented:
> Keyboard UI: If you shift+tab to the stream input, it should open without making you hit Enter. As it is, there's no indication that you've tabbed to the right place, which is bad, and also if you start typing once moving focus to the (closed) dropdown_list_widget, it triggers keyboard shortcuts and more or less randomly navigates you to a different part of the app. It's possible this could be done as a general on-focus handler for the dropdown_list_widget outer button; if so please just add it as a new commit on top.

I don't use keyboard shortcuts enough to have a good intuition for how this should work, so I want to clarify some things.

* from topic input, shift+tab to stream input opens the dropdown with focus in the dropdown's search bar
* once the dropdown open
  * if the user starts typing, that text shows up in the dropdown search bar
  * the user must **press tab twice** to go back to the topic input (the first tab focuses on the results of the search)

Does that seem reasonable?

sahil839 commented:
> Correct, this is I think pretty important. The "Move topic" modal would benefit from this as well, so this can just be a generic dropdown_list_widget feature; mentioning @sahil839 in case he has time to pick up just implementing this detail, since I know you have a lot on your plate.

This means we will open the dropdown without pressing `Enter`. This is not what happens with the `select` element dropdowns, they are opened on pressing `Enter` when they are in focus. Do we want to do the same for `dropdown_list_widget`? I don't think a user expects anything other than focussing the input/button when using `Tab` or `Shift+Tab`. 

> I'd usually expect to need to do something like Esc or select an option to close this before tabbing elsewhere

I also noticed that the menu does not close on pressing `Esc` currently, And for the "Move topic" modal, the modal itself on pressing `Esc` when dropdown menu is opened.

evykassirer commented:
Started a conversation on CZO about this: https://chat.zulip.org/#narrow/stream/6-frontend/topic/dropdown.20keyboard.20behavior

evykassirer commented:
Just pushed a fix for the "composebox closing on click" issue.

evykassirer commented:
> I also noticed that the menu does not close on pressing Esc currently

Maybe I'm not reproducing correctly, but it seems like this is fixed now

evykassirer commented:
Keyboard navigation (not having to press enter) is the one thing I haven't addressed yet that I know of.

evykassirer commented:
Oh I guess the other thing is I haven't added support for short screens yet, but I'm not sure if that's a blocker here.

timabbott commented:
Test-deployed this on chat.zulip.org.

evykassirer commented:
changed: unit tests pass now.

remaining:

* Keyboard navigation (not having to press enter)
* Short screen support

alya commented:
@sahil839 I wonder if you'd be up for taking over this part of the changes, since @evykassirer is juggling a lot of PR feedback at the moment?

> Keyboard navigation (not having to press enter)

sahil839 commented:
> @sahil839 I wonder if you'd be up for taking over this part of the changes, since @evykassirer is juggling a lot of PR feedback at the moment?
> 
> > Keyboard navigation (not having to press enter)

Yeah sure.

sahil839 commented on web/src/ui_init.js:
```diff
@@ -653,6 +652,7 @@ export function initialize_everything() {
     message_view_header.initialize();
     server_events.initialize();
     user_status.initialize(user_status_params);
+    compose_recipient.initialize();
```
Why this is initialized twice? I am assuming it is by mistake, but just want to confirm. This is causing problems in implementing the keyboard behavior to open on focus.

evykassirer replied:
Oh yup a mistake, thanks for catching that!

evykassirer replied:
Fixed in most recent psuh

sahil839 commented on web/src/compose_ui.js:
```diff
@@ -500,3 +511,49 @@ export function get_submit_button() {
+        .filter((stream) => stream_data.can_post_messages_in_stream(stream))
+        .map((stream) => ({
+            name: stream.name,
+            value: stream.name,
```
You can update this object to add `stream` field to add the stream icon to the dropdown to make it consistent with othe places. You can check code in `settings_org.init_dropdown_widgets` for reference.

evykassirer replied:
Oh cool!! Thanks :)

evykassirer replied:
Switched to this in the most recent push, way cleaner :) Thanks for pointing it out!

sahil839 commented:
> > @sahil839 I wonder if you'd be up for taking over this part of the changes, since @evykassirer is juggling a lot of PR feedback at the moment?
> > > Keyboard navigation (not having to press enter)
> 
> Yeah sure.

This seems somewhat complex to implement. While adding code to open on "focus" event can be done easily, but this breaks the click behavior. Specifically, when the menu is opened, clicking on the button should close it. But this is what happens, when I click on the button when the dropdown is opened -
- The focus event is triggered due to click. But nothing happens, because the code added in `focus` event handler only opens the menu if it is closed previously.
- Then click event is triggered, which is handled by bootstrap, and this closes the menu.
- The `click` handler in bootstrap again calls focus as per [code here](https://github.com/zulip/zulip/blob/c995e48e11c195698a755049e4f5084349827f5b/web/third/bootstrap/js/bootstrap.js#L370) and this then opens the menu again as it was previously closed.

sahil839 commented:
I added the commit to fix the keyboard behavior. @evykassirer FYI, would be good if you could test it once.

timabbott commented on web/src/hotkey.js:
```diff
@@ -582,6 +583,11 @@ export function process_shift_tab_key() {
+    if ($("#stream_message_recipient_topic").is(":focus")) {
+        compose_recipient.open_compose_stream_dropup();
+        return true;
+    }
```
This feels great! We may want to do a similar check in the "Move topic" modal, since we start focus in the topic input there too, but that can be a follow-up.

timabbott commented on templates/zerver/emails/compiled/digest.html:
```diff
@@ -0,0 +1,43 @@
+<p style="font-family:sans-serif; font-size:14px; font-weight:normal; margin:10px 0">
+    <a href="{{ realm_uri }}/#settings/notifications" style="color:#5f5ec7; text-decoration:underline">{% trans %}Manage email preferences{% endtrans %}</a> |
+    <a href="{{ unsubscribe_link }}" style="color:#5f5ec7; text-decoration:underline">{% trans %}Unsubscribe from digest emails{% endtrans %}</a>
+</p>
```
Looks like you accidentally applied `git add` to these files that we'd removed.

timabbott replied:
I fixed this -- was the cause of CI failures.

timabbott commented:
I'm opening issues to track the remaining work for this:
* https://github.com/zulip/zulip/issues/25217
* https://github.com/zulip/zulip/issues/25218
* https://chat.zulip.org/#narrow/stream/101-design/topic/read-only.20streams.20woes/near/1552131 is probably a blocker for merging this, unfortunately.

timabbott commented:
Merged, thanks for all the work on this @sahil839 and @evykassirer!
